block/block_service.proto

```diff
@@ -0,0 +1,710 @@
+        /**
+         * An "unset value" flag, this value SHALL NOT be used.<br/>
+         * This status indicates the server software failed to set a
+         * status, and SHALL be considered a software defect.
+         */
+        STREAM_ITEMS_UNKNOWN = 0;
+
+        /**
+         * The request succeeded.<br/>
+         * No errors occurred and the source node orderly ended the stream.
+         */
+        STREAM_ITEMS_SUCCESS = 1;
+
+        /**
+         * The delay between items was too long.<br/>
+         * The source MUST start a new stream before the failed block.
+         */
+        STREAM_ITEMS_TIMEOUT = 2;
+
+        /**
+         * An item was received out-of-order.<br/>
+         * The source MUST start a new stream before the failed block.
+         */
+        STREAM_ITEMS_OUT_OF_ORDER = 3;
+
+        /**
+         * A block state proof item could not be validated.<br/>
+         * The source MUST start a new stream before the failed block.
+         */
+        STREAM_ITEMS_BAD_STATE_PROOF = 4;
+    }
+}
+
+// ---- ReadBlock RPC ----
+
+/**
+ * A request to read a single block.
+ *
+ * A client system SHALL send this message to request a single block,
+ * including the block state proof.
+ *
+ * > REVIEW NOTE
+ * >> We _could_ make `block_number` optional and remove `retrieve_latest`.
+ * >> An **unset** `block_number` would then result in returning the latest
+ * >> block.<br/>
+ * >> It is not entirely clear which option would be easier for developers to
+ * >> use, but initial intuition is that a boolean is easier to understand.
+ */
+message SingleBlockRequest {
+    /**
+     * The block number of a block to retrieve.
+     * <p>
+     * The requested block MUST exist on the block node.<br/>
+     * This value MUST NOT be set if `retrieve_latest` is set `true`.<br/>
+     * This value MUST be set to a valid block number if `retrieve_latest` is
+     * unset or is set `false`.
+     */
+    uint64 block_number = 1;
+
+    /**
+     * A boolean to request the latest available block.
+     * <p>
+     * This value MAY be set `true` to request the last block available.<br/>
+     * If this value is set to `true` then `block_number` MUST NOT be set and
+     * SHALL be ignored.
+     */
+    bool retrieve_latest = 3;
+}
+
+/**
+ * A response to a `readBlock` request.
+ *
+ * This message SHALL be sent in response to a request, and SHALL contain at
+ * least a valid `status`.<br/>
+ * If `status` is `READ_BLOCK_SUCCESS`, the response SHALL contain the
+ * requested block in the `block` field.
+ *
+ * > Note
+ * >> A block can become quite large. A client MUST be prepared to receive the
+ * >> full content of the block, perhaps many megabytes of data.
+ */
+message SingleBlockResponse {
+    /**
+     * A response status.
+     * <p>
+     * The reported status SHALL reflect the success of the request, or
+     * a detailed reason the request failed.
+     */
+    SingleBlockResponse.ResponseCode status = 1;
```

Comment:
In `SubscribeStreamResponse` you had a `oneof` here.  Should the two rsponses be similar in this case - either a `oneof` or a separation of fields

Reply:
Stream responses are sent _many_ times for a single request, and there are a couple different response types possible because we need to send data items and result codes, but cannot joint the two together (they're sent at different times).

The Single block response is sent exactly once per request, and has only one possible response type which includes both data and result code.